### PR TITLE
Moving the cloudprovider interface to staging.

### PR DESCRIPTION
**Pull request by cheftako**
**What this PR does / why we need it**:
Individual implementations are not yet being moved.
Fixed all dependencies which call the interface.
Fixed golint exceptions to reflect the move.
One of the first steps for https://github.com/kubernetes/community/blob/master/keps/sig-cloud-provider/0013-build-deploy-ccm.md

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # KEP-0013

**Special notes for your reviewer**:

**Release note**:
```release-note
Any external provider should be aware the cloud-provider interface should be imported from :-
cloudprovider "k8s.io/cloud-provider"
```

### CODE REVIEW

File: hack/.golint_failures [hunk 1]
```diff
@@ -686,6 +685,7 @@ staging/src/k8s.io/client-go/util/jsonpath
 staging/src/k8s.io/client-go/util/retry
 staging/src/k8s.io/client-go/util/testing
 staging/src/k8s.io/client-go/util/workqueue
+staging/src/k8s.co/cloud-provider
```
Comment (andrewsykim) [inline hack/.golint_failures]: s/k8s.co/k8s.io

Comment (cheftako) [reply hack/.golint_failures]: LOL. Thank you :D

**Comment by andrewsykim** [comment]
cc @d-nishi @mcrute

File: pkg/cloudprovider/doc.go [hunk 2]
```diff
@@ -15,4 +15,5 @@ limitations under the License.
 */
 
 // Package cloudprovider supplies interfaces and implementations for cloud service providers.
-package cloudprovider // import "k8s.io/kubernetes/pkg/cloudprovider"
+// To use: import cloudprovider "k8s.io/cloud-provider"
```
Comment (justinsb) [inline pkg/cloudprovider/doc.go]: Too late to change to `k8s.io/cloudprovider` ?

File: staging/src/BUILD [hunk 3]
```diff
@@ -189,6 +189,7 @@ filegroup(
         "//staging/src/k8s.io/client-go/util/retry:all-srcs",
         "//staging/src/k8s.io/client-go/util/testing:all-srcs",
         "//staging/src/k8s.io/client-go/util/workqueue:all-srcs",
+        "//staging/src/k8s.io/cloud-provider:all-srcs",
```
Comment (justinsb) [inline staging/src/BUILD]: Can't comment on the `100755 → 100644` mode changes, but do they belong in this PR?  Maybe send a separate PR?

File: staging/src/k8s.io/cloud-provider/OWNERS [hunk 4]
```diff
@@ -0,0 +1,48 @@
+approvers:
+- mikedanese
+- dims
+- wlan0
+- andrewsykim
+- cheftako
+reviewers:
+- thockin
```
Comment (justinsb) [inline staging/src/k8s.io/cloud-provider/OWNERS]: Might be worth trimming this list for faster reviews, but :man_shrugging:

File: staging/src/k8s.io/cloud-provider/README.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+**Note:** go-get or vendor this package as `k8s.io/cloud-provider`.
+
+## Purpose
+
+This library is a shared dependency for processes which need to be able to
+integrate with cloud-provider specific functionality.
+
+## Compatibility
+
+Cloud Providers are expected to keep their implementations in sync with the
```
Comment (justinsb) [inline staging/src/k8s.io/cloud-provider/README.md]: Perhaps "...the HEAD of their implementations in sync..."

**Review by justinsb — COMMENTED** [review]
Some minor suggestions, and raising the question of "cloud-provider" vs "cloudprovider" as the alias import is awkward, but LGTM

**Comment by mcrute** [comment]
This looks good from the AWS side.

File: pkg/cloudprovider/doc.go (hunk 2 above)
Comment (cheftako) [reply pkg/cloudprovider/doc.go]: Its not too late yet. I agree its awkward but it is also a common pattern we have across all of our staging repos. The problem is exacerbated because its a shallow repo. I had deliberately kept the common pattern but if we think is not a pattern we want to keep I can change it.

File: staging/src/BUILD (hunk 3 above)
Comment (cheftako) [reply staging/src/BUILD]: I agree they do not belong in this PR. Will make the change.

File: staging/src/k8s.io/cloud-provider/OWNERS (hunk 4 above)
Comment (cheftako) [reply staging/src/k8s.io/cloud-provider/OWNERS]: Good point.

File: staging/src/k8s.io/cloud-provider/README.md (hunk 5 above)
Comment (cheftako) [reply staging/src/k8s.io/cloud-provider/README.md]: Done

**Comment by justinsb** [comment]
Looks like you pulled the file mode changes out (thank you) and lost your lgtm because of it.

/lgtm

**Comment by dims** [comment]
/approve
/lgtm

Looks good! thanks @cheftako

File: cmd/kube-controller-manager/app/plugins.go [hunk 6]
```diff
@@ -26,7 +26,7 @@ import (
 	"github.com/golang/glog"
 
 	// Cloud providers
-	"k8s.io/kubernetes/pkg/cloudprovider"
+	cloudprovider "k8s.io/cloud-provider"
```
Comment (lavalamp) [inline cmd/kube-controller-manager/app/plugins.go]: Is the "-" worth it? :)

File: pkg/cloudprovider/doc.go (hunk 2 above)
Comment (lavalamp) [reply pkg/cloudprovider/doc.go]: I think it'd be better to change to one of:
* k8s.io/cloudprovider
* k8s.io/cloud-provider/cloudprovider

But I'm not 100% sure it is worth the effort if this is ready to go.

File: staging/src/k8s.io/cloud-provider/README.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+**Note:** go-get or vendor this package as `k8s.io/cloud-provider`.
+
+## Purpose
+
+This library is a shared dependency for processes which need to be able to
+integrate with cloud-provider specific functionality.
+
+## Compatibility
+
+Cloud Providers are expected to keep the HEAD of their implementations in sync
```
Comment (lavalamp) [inline staging/src/k8s.io/cloud-provider/README.md]: This doesn't sound like a reasonable request?

My (untested) goals for the apply repo look like this: https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/RELEASE.md

File: staging/src/k8s.io/cloud-provider/README.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+## Where does it come from?
+
+`cloud-provider` is synced from
+https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider.
+Code changes are made in that location, merged into k8s.io/kubernetes and
+later synced here.
+
+## Things you should NOT do
+
+ 1. Add an cloud provider specific code to this repo.
```
Comment (lavalamp) [inline staging/src/k8s.io/cloud-provider/README.md]: s/an/any/

**Comment by lavalamp** [comment]
/approve

I made some comments on the repo setup .md files; you can address in a followup. (Next time consider sending those first in a separate PR? Not that there should be a next time...)

Also I'll leave a hold in case you think it's worth removing the '-' from the package name; just unhold if not.

/hold

**Comment by dims** [comment]
i am +1 to keeping the `-`

**Comment by hogepodge** [comment]
Also +1 on consistency. /lgtm

**Comment by lavalamp** [comment]
OK in that case:

/hold cancel
/unhold
